counter.go

```diff
@@ -0,0 +1,127 @@
+package signalfx
+
+import (
+	"math"
+	"sync/atomic"
+	"time"
+)
+
+// A Counter represents a Counter metric (i.e., its internal value is
+// reset to zero by its PostReportHook, although this should not be
+// meaningful to client code).
+type Counter struct {
+	metric     string
+	dimensions map[string]string
+	value      uint64
+}
+
+// NewCounter returns a new Counter with the specified parameters.  It
+// does not copy the dimensions; client code should take care not to
+// modify them in a goroutine-unsafe manner.
+func NewCounter(
+	metric string,
+	dimensions map[string]string,
+	value uint64,
+) *Counter {
+	return &Counter{metric: metric, dimensions: dimensions, value: value}
+}
+
+// Inc increments a Counter's internal state.  It is goroutine-safe.
+func (c *Counter) Inc(delta uint64) uint64 {
+	return atomic.AddUint64(&c.value, delta)
+}
+
+// DataPoint returns a DataPoint reflecting the Counter's internal
+// state.  It does not reset that state, leaving that to
+// PostReportHook.
+func (c *Counter) DataPoint() *DataPoint {
+	value := atomic.LoadUint64(&c.value)
+	if value == 0 || value > math.MaxInt64 {
+		return nil
+	}
+	return &DataPoint{
+		Metric:     c.metric,
+		Timestamp:  time.Now(),
+		Type:       CounterType,
+		Dimensions: c.dimensions,
+		Value:      int64(value),
+	}
+}
+
+// PostReportHook resets a Counter's internal state.  It does this
+// intelligently, by subtracting the successfully-reported value from
+// the state (this way, the Counter may have been incremented in
+// between the create of the reported DataPoint and the reset, and the
+// state will be consistent).  Calling PostReportHook with a negative
+// value will result in a panic: counters may never take negative
+// values.
+func (c *Counter) PostReportHook(v int64) {
+	if v < 0 {
+		panic("negative counter should be impossible")
```

Comment:
Good catch @rajnmithun. My rule of thumb is to panic for things that are developer error, e.g. if you create a template that doesn't compile, it should panic at the start of execution (http://golang.org/pkg/text/template/#Must). However, it is probably not a good idea to panic for weird runtime situations, because what we think is impossible as developers has a way of becoming possible in production.

Sure it is possible to recover from panic, but I doubt that people using this package would know that they need to prepare for possible panics. As a result, this could crash production systems which is definitely a blocking issue.